File: src/main/java/io/github/slimefunguguproject/bump/implementation/listeners/BowShootListener.java
package io.github.slimefunguguproject.bump.implementation.listeners;

import org.bukkit.entity.Arrow;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.EntityShootBowEvent;

import io.github.slimefunguguproject.bump.core.handlers.BowUseHandler;
import io.github.thebusybiscuit.slimefun4.api.items.SlimefunItem;

/**
 * This {@link Listener} will call {@link BowUseHandler}
 * when {@link Player} fires with a bow.
 *
 * @author ybw0014
 * @see BowUseHandler
 */
public final class BowShootListener implements Listener {
    @EventHandler
    public void onBowUse(EntityShootBowEvent e) {
        if (e.getEntity() instanceof Player player && e.getProjectile() instanceof Arrow) {
            SlimefunItem bow = SlimefunItem.getByItem(e.getBow());

            if (bow != null) {
                bow.callItemHandler(BowUseHandler.class, handler -> handler.onUse(e, player, e.getBow()));
            }
        }
    }
}

File: src/main/kotlin/io/github/slimefunguguproject/bump/implementation/listeners/BowShootListener.kt
package io.github.slimefunguguproject.bump.implementation.listeners

import io.github.slimefunguguproject.bump.Bump
import io.github.slimefunguguproject.bump.core.handlers.BowUseHandler
import io.github.thebusybiscuit.slimefun4.api.items.SlimefunItem
import org.bukkit.entity.Arrow
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.entity.EntityShootBowEvent

class BowShootListener(plugin: Bump) : Listener {
    init {
        plugin.server.pluginManager.registerEvents(this, plugin)
    }

    @EventHandler
    fun onBowUse(e: EntityShootBowEvent) {
        if (e.entity is Player && e.projectile is Arrow) {
            val bow = SlimefunItem.getByItem(e.bow)

            bow?.callItemHandler(BowUseHandler::class.java) { handler: BowUseHandler ->
                handler.onUse(e, e.entity as Player, e.bow!!)
            }
        }
    }
}
